File: src/main/java/ru/melonhell/bulletphysics/utils/math/BetweenClosedUtils.java
package ru.melonhell.bulletphysics.utils.math;

import com.jme3.bounding.BoundingBox;
import com.jme3.math.Vector3f;
import lombok.experimental.UtilityClass;
import ru.melonhell.bulletphysics.nms.wrappers.BlockPos;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

@UtilityClass
public class BetweenClosedUtils {
    public List<BlockPos> betweenClosed(BoundingBox box) {
        List<BlockPos> result = new ArrayList<>();
        betweenClosed(box, result);
        return result;
    }

    public void betweenClosed(BoundingBox box, Collection<BlockPos> result) {
        Vector3f min = box.getMin(null);
        Vector3f max = box.getMax(null);
        betweenClosed((int) Math.floor(min.x), (int) Math.floor(min.y), (int) Math.floor(min.z), (int) Math.floor(max.x), (int) Math.floor(max.y), (int) Math.floor(max.z), result);
    }

    public List<BlockPos> betweenClosed(int startX, int startY, int startZ, int endX, int endY, int endZ) {
        List<BlockPos> result = new ArrayList<>();
        betweenClosed(startX, startY, startZ, endX, endY, endZ, result);
        return result;
    }

    public void betweenClosed(int startX, int startY, int startZ, int endX, int endY, int endZ, Collection<BlockPos> result) {
        for (int x = startX; x < endX; x++) {
            for (int y = startY; y < endY; y++) {
                for (int z = startZ; z < endZ; z++) {
                    result.add(new BlockPos(x, y, z));
                }
            }
        }
    }
}

File: src/main/java/ru/melonhell/bulletphysics/utils/math/BetweenClosedUtils.kt
package ru.melonhell.bulletphysics.utils.math

import com.jme3.bounding.BoundingBox
import ru.melonhell.bulletphysics.nms.wrappers.BlockPos

object BetweenClosedUtils {
    fun betweenClosed(box: BoundingBox): List<BlockPos> {
        val result: MutableList<BlockPos> = ArrayList()
        betweenClosed(box, result)
        return result
    }

    fun betweenClosed(box: BoundingBox, result: MutableCollection<BlockPos>) {
        val min = box.getMin(null)
        val max = box.getMax(null)
        betweenClosed(
            Math.floor(min.x.toDouble()).toInt(),
            Math.floor(min.y.toDouble()).toInt(),
            Math.floor(min.z.toDouble()).toInt(),
            Math.floor(max.x.toDouble()).toInt(),
            Math.floor(max.y.toDouble()).toInt(),
            Math.floor(max.z.toDouble()).toInt(),
            result
        )
    }

    fun betweenClosed(startX: Int, startY: Int, startZ: Int, endX: Int, endY: Int, endZ: Int): List<BlockPos> {
        val result: MutableList<BlockPos> = ArrayList()
        betweenClosed(startX, startY, startZ, endX, endY, endZ, result)
        return result
    }

    fun betweenClosed(
        startX: Int,
        startY: Int,
        startZ: Int,
        endX: Int,
        endY: Int,
        endZ: Int,
        result: MutableCollection<BlockPos>
    ) {
        for (x in startX until endX) {
            for (y in startY until endY) {
                for (z in startZ until endZ) {
                    result.add(BlockPos(x, y, z))
                }
            }
        }
    }
}